server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -734,6 +737,8 @@ public IndexMetadata withInSyncAllocationIds(int shardId, Set<String> inSyncSet)
         if (inSyncSet.equals(inSyncAllocationIds.get(shardId))) {
             return this;
         }
+        List<Set<String>> copiedInSyncAllocationIds = new ArrayList<>(this.inSyncAllocationIds);
+        copiedInSyncAllocationIds.set(shardId, Set.copyOf(inSyncSet));
```

Comment:
I feel that we need to check the invariant that `shardId` is lower than `inSyncAllocationIds`'s size, otherwise the `set` call would fail. Or does `copiedInSyncAllocationIds` need to resized if `shardId` is bigger than `inSyncAllocationIds`?

Reply:
The copied inSyncAllocationIds is an arraylist that is initialized with numberOfShards size. So the invariant should hold and I can add an assertion here (if it is sufficient to check the invariant like that).